[Compose 페이먼츠 1단계] 다이스 미션 제출합니다.

moondev03 opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
```
이번 리뷰 잘 부탁드립니다 🙇‍♂️
```

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

[record.webm](https://github.com/user-attachments/assets/ca06541c-f78d-4907-8be3-1ab464bae09f)

moondev03 committed:
docs: 기능 요구 사항 작성

moondev03 committed:
build: compose 테스트 의존성 추가

moondev03 committed:
feat: 카드 번호 입력창 구현

moondev03 committed:
feat: 카드 비밀번호 입력창 구현

yrsel committed:
style: ktlint compose ruleset 추가

moondev03 committed:
feat: 카드 만료일 입력창 구현

moondev03 committed:
feat: 카드 소유자 이름 입력창 구현

moondev03 committed:
feat: 카드 등록 화면 TopBar 구현

moondev03 committed:
feat: 카드 미리보기 컴포넌트 구현

moondev03 committed:
build: kotlin-parcelize 플러그인 추가

moondev03 committed:
feat: 카드 유효성 검증 로직 구현

moondev03 committed:
chore: Ktlint Format 수행

moondev03 committed:
feat: 카드 등록 화면 UI 상태 정의

moondev03 committed:
feat: 카드 등록 화면 구현

- 카드 등록 화면을 구성하고 TopAppBar를 추가합니다.
- 카드 번호, 만료일, 소유자 이름, 비밀번호 입력 필드를 통합합니다.
- 입력 값 유효성 검사 및 스낵바 메시지 기능을 추가합니다.
- 만료된 카드에 대한 메시지를 문자열 리소스에 추가합니다.

moondev03 committed:
test: 카드 등록 버튼 활성화 테스트 추가

moondev03 committed:
docs: README.md 완료 항목 표시

moondev03 committed:
refactor: MaskVisualTransformation 삭제 및 PasswordVisualTransformation 적용

moondev03 committed:
refactor: CardRegistrationScreenUiState 상태 관리 로직 개선

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/common/ExpirationDateVisualTransformation.kt:
```diff
@@ -0,0 +1,40 @@
+    private const val EXPIRATION_DATE_LENGTH = 4
+    private const val EXPIRATION_DATE_SEPARATOR = " / "
+
+    private val offsetMapping =
```
`CreditCardVisualTransformation`에서와는 다르게 여기에서는 프로퍼티로 선언해주셨습니다. 어떤 차이가 있을까요? 🤔

moondev03 replied:
OffsetMapping를 매번 생성하지 않도록 하기 위함이었는데 성능상 큰 차이가 없던 것 같습니다.
추가로 그룹으로 나눠 구분자를 추가하는 로직은 현재 시점에서는 모두 동일한 기능인 것 같아서 GroupedVisualTransformation으로 통합하였습니다.

[4b4d1ab](https://github.com/woowacourse/android-payments/pull/25/commits/4b4d1ab7d29e05791bb4f7ae51483f0c618d1aa7)

giovannijunseokim replied:
`GroupedVisualTransformation` 클래스를 생성하고, 잘 재사용 해주셨네요! 💯

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/common/ExpirationDateVisualTransformation.kt:
```diff
@@ -0,0 +1,40 @@
+            override fun originalToTransformed(offset: Int): Int =
+                when (offset) {
+                    in 1..2 -> offset
+                    in 3..4 -> offset + 3
```
여기 `offset + 3`의 `3`은 무슨 의미일까요? 👀

moondev03 replied:
[4b4d1ab](https://github.com/woowacourse/android-payments/pull/25/commits/4b4d1ab7d29e05791bb4f7ae51483f0c618d1aa7)

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/common/ExpirationDateVisualTransformation.kt:
```diff
@@ -0,0 +1,40 @@
+
+            override fun transformedToOriginal(offset: Int): Int {
+                val group = offset / 5
+                val posInGroup = offset % 5
```
`posInGroup`이라는 변수명이 의미하는 것이 궁금해요! 🧐

moondev03 replied:
[4b4d1ab](https://github.com/woowacourse/android-payments/pull/25/commits/4b4d1ab7d29e05791bb4f7ae51483f0c618d1aa7)

giovannijunseokim replied:
조금 더 정확히 말씀드리면 `pos`라는 단어의 의미를 모르겠어요. `position`의 약자인가요? 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/payments/CardExpirationDateTextField.kt:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    OutlinedTextField(
+        modifier = modifier.testTag(CARD_EXPIRATION_DATE_TEXT_FIELD_TEST_TAG),
```
테스트를 위한 코드가 프로덕션 코드가 반영되어도 괜찮을까요? 🤔

moondev03 replied:
테스트 코드에서 testTag를 붙인 Modifier를 주입하도록 할 수 있겠군요!
수정했습니다 :)

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/payments/CardExpirationDateTextField.kt:
```diff
@@ -0,0 +1,95 @@
+    cardExpirationDate: String,
+    onCardExpirationDateChanged: (String) -> Unit,
+    errorMessage: String? = null,
+    onErrorMessageChanged: (String?) -> Unit,
```
기본 인자가 있는 `modifier` 파라미터가 맨 앞에 와도 괜찮을까요? 🤔

moondev03 replied:
찾아보니 안드로이드 컴포즈 컴포넌트 가이드라인에서는 아래와 같은 이유로 첫번째 선택적 매개 변수로 하는게 권장되는 것 같네요!

d8d81ac

```
- 구성 요소가 0이 아닌 기본 크기를 가지고 있다면 구성 요소는 자급자족하기 때문에 수정자는 선택 사항이어야 합니다. 
- 수정자는 모든 구성 요소에 권장되며 자주 사용되기 때문에 먼저 배치하면 명명된 매개 변수 없이 설정할 수 있으며 모든 구성 요소에서 이 매개 변수에 대한 일관된 위치를 제공합니다.
```
[참고한 문서](https://github.com/androidx/androidx/blob/androidx-main/compose/docs/compose-component-api-guidelines.md#modifier-parameter)

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/payments/CardNumberTextField.kt:
```diff
@@ -0,0 +1,59 @@
+            }
+            if (newValue.isDigitsOnly().not()) return@OutlinedTextField
+            onCardNumberChanged(newValue)
+        },
```
조금 더 간결하게 작성할 수 있을까요? 👀 
코드 순서가 조금 바뀌면 가능할지도 모르겠어요.

moondev03 replied:
모든 컴포넌트 수정했습니다!

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/payments/CardRegistrationScreen.kt:
```diff
@@ -0,0 +1,155 @@
+    paymentCardValidator.validateCardNumber(cardNumber) &&
+        paymentCardValidator.validateCardExpirationDate(cardExpirationDate) &&
+        paymentCardValidator.validateCardholderName(cardholderName) &&
+        paymentCardValidator.validateCardPassword(cardPassword)
```
이 함수는 `CardRegistrationScreenUiState`의 멤버 함수가 될 수도 있을 것 같은데요. 이 곳에 있는 이유는 무엇일까요? 🤔

giovannijunseokim replied:
또한 이 함수의 반환값은 `CardRegistrationScreenUiState`의 프로퍼티 `isRegistrableCard`와 어떤 차이가 있을까요? 👀

moondev03 replied:
해당 함수의 반환값를 바로 사용하거나 UiState 내부에서 처리하는게 더 적절해 보이네요!

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/payments/CardRegistrationScreen.kt:
```diff
@@ -0,0 +1,155 @@
+            cardExpirationDate = uiState.cardExpirationDate,
+            onCardExpirationDateChanged = { newValue ->
+                val isValid =
+                    newValue.length != 4 || paymentCardValidator.validateCardExpirationDate(newValue)
```
새롭게 입력된 값의 길이를 검증하는 것은 누구의 책임이라고 생각하시나요? 👀

moondev03 replied:
원래 길이에 대한 검증을 CardExpirationDateTextField가 스스로 판단하도록 했습니다.
다만 여기서는 유효성 검사를 위해 4글자일 때만 검증을 하도록 하기 위함이었어요

현재는 수정했습니다 :)

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/payments/CardRegistrationScreenUiState.kt:
```diff
@@ -0,0 +1,15 @@
+    val cardPassword: String = "",
+    val isRegistrableCard: Boolean = false,
+    val snackbarMessage: String? = null,
+) : Parcelable
```
`UiState`가 `Parcelable`을 구현하는 이유는 무엇인가요? 👀

moondev03 replied:
아래와 같이 `rememberSaveable`을 사용하면,

```kotlin
var uiState by rememberSaveable { mutableStateOf(CardRegistrationScreenUiState()) }
```

`CardRegistrationScreenUiState`가 단순 `data class`일 경우 다음과 같은 에러가 발생합니다.

```
MutableState containing CardRegistrationScreenUiState(...) cannot be saved using the current SaveableStateRegistry.
The default implementation only supports types which can be stored inside the Bundle.
```

이는 `rememberSaveable`이 내부적으로 `Bundle`을 통해 상태를 저장하기 때문에 `Bundle`에 저장 가능한 타입만 기본적으로 지원하기 때문이라고 합니다.

---

이를 해결하기 위한 방법으로는 2가지 방안이 있었습니다.
이 중 간단하게 사용가능한 Parcelable 방식을 선택했습니다.

1. **`Parcelable` 적용**  
   `@Parcelize`를 사용해 `CardRegistrationScreenUiState`를 `Parcelable`로 만들면 별도 설정 없이 `rememberSaveable`에 바로 저장할 수 있습니다.

   ```kotlin
   @Parcelize
   data class CardRegistrationScreenUiState(
       val cardNumber: String = "",
       val cardExpirationDate: String = "",
       val cardExpirationDateErrorMessage: String? = null,
       val cardholderName: String = "",
       val cardPassword: String = "",
   ) : Parcelable
   ```

2. **`Saver` 직접 구현**  
   `Saver`를 정의해 직렬화/복원 방식을 명시적으로 제어할 수도 있습니다.

   ```kotlin
   val CardRegistrationScreenUiStateSaver = Saver<CardRegistrationScreenUiState, Any>(
       save = { state ->
           listOf(
               state.cardNumber,
               state.cardExpirationDate,
               state.cardExpirationDateErrorMessage,
               state.cardholderName,
               state.cardPassword
           )
       },
       restore = { restored ->
           val list = restored as List<*>
           CardRegistrationScreenUiState(
               cardNumber = list[0] as String,
               cardExpirationDate = list[1] as String,
               cardExpirationDateErrorMessage = list[2] as String?,
               cardholderName = list[3] as String,
               cardPassword = list[4] as String
           )
       }
   )

   var uiState by rememberSaveable(stateSaver = CardRegistrationScreenUiStateSaver) {
       mutableStateOf(CardRegistrationScreenUiState())
   }
   ```

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/payments/CardholderNameTextField.kt:
```diff
@@ -0,0 +1,76 @@
+                modifier = Modifier.fillMaxWidth(),
+            )
+        },
+        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Text),
```
현재 이 `TextField`는 영어만 입력받도록 되어있는 것 같아요. 더욱 적절한 `KeyboardType`은 없을까요? 👀

moondev03 replied:
`KeyboardType.Ascii`가 있었네요 :)

aeac4d2

giovannijunseokim commented on app/src/main/java/woowacourse/payments/util/PaymentCardValidator.kt:
```diff
@@ -0,0 +1,11 @@
+package woowacourse.payments.util
+
+interface PaymentCardValidator {
```
이 추상화의 이점은 무엇인가요? 🤔

moondev03 replied:
카드의 유효성 검사의 조건이 변경될 수 있다고 생각해서 검증을 위한 객체를 추상화하였습니다.

giovannijunseokim replied:
현재는 `PaymentCardValidator`의 구현체로는 `DefaultPaymentCardValidator`만 존재하고 있는 것 같아요.

추후 다른 방식으로 결제 카드 정보를 검증하는 방식이 등장할 수 있다고 본 것일까요?
아니면 해당 객체에 의존하고 있는 객체를 테스트하기 위함인가요? 👀 

> 카드의 유효성 검사의 조건이 변경될 수 있다고 생각해서 검증을 위한 객체를 추상화하였습니다.

변경될 수 있는 모든 것에는 추상화가 필요한가요? 다이스가 어떤 객체 또는 개념을 추상화하는 조건이 궁금합니다. 🤔

giovannijunseokim reviewed (changes requested):
안녕하세요 다이스. 미션 구현 고생하셨습니다! 👍 

궁금한 점들을 리뷰 내용으로 남겨두었으니 확인 후 답변 부탁드립니다. 😁 

궁금한 점이나 논의하고 싶은 내용이 있으면 편하게 얘기해주세요! 🙏

moondev03 committed:
style: Modifier 매개변수 순서 변경

moondev03 committed:
refactor: 카드 소유자 이름 키보드 타입 변경

moondev03 committed:
refactor: VisualTransformation을 일반화된 GroupedVisualTransformation으로 변경

moondev03 committed:
test: compose test 코드 명시적 파라미터명 변경

moondev03 committed:
chore: compose test 코드 정리

- compose test 코드에서 modifier에 testTag를 주입하도록 변경
- preview 코드 정리

moondev03 committed:
refactor: 카드 만료일 입력창 유효성 관련 상태 관리 개선

moondev03 committed:
refactor: 카드 소유자 입력창 코드 개선

moondev03 committed:
refactor: 카드 번호 입력창 코드 개선

moondev03 committed:
refactor: 카드 비밀번호 입력창 코드 개선

moondev03 committed:
refactor: 카드 만료일 입력값 길이 제한

moondev03 committed:
refactor: 카드 소유자 입력값 길이 제한

moondev03 committed:
refactor: 카드 번호 입력값 길이 제한

moondev03 committed:
feat: 입력창 오류 메시지 표시 기능 추가

moondev03 committed:
refactor: 카드 유효성 검사 모듈 도메인 레이어로 이동

moondev03 committed:
refactor: ui 컴포넌트 패키지 변경

- `payments` 패키지에서 `component` 패키지로 이동

moondev03 committed:
feat: StringResWithParams 구현

- 리소스 id와 파라미터를 받아 문자열을 반환하는 StringResWithParams 클래스를 구현했습니다.

moondev03 committed:
refactor: 카드 정보 유효성 검사 결과 세분화

카드 정보 유효성 검사 결과를 기존 Boolean에서
`NOT_FILLED`, `INVALID`, `VALID` 세 가지 상태를 가지는
`PaymentCardValidationResult` enum class로 변경했습니다.

moondev03 committed:
refactor: 카드 등록 화면 패키지 이동

- 카드 등록 화면 관련 파일을 `ui.payments`에서 `ui.screen.registration` 패키지로 이동합니다.

moondev03 committed:
fix: 유효기간 입력창 오류 메시지 로직 수정

- 에러 메시지 우선 순위 변경

moondev03 committed:
refactor: CardRegistrationScreen을 ViewModel을 이용하도록 리팩토링

- CardRegistrationScreenViewModel 생성
- CardRegistrationScreenUiState, CardRegistrationScreenUiEvent 생성
- CardRegistrationScreenViewModelSaver를 이용해 rememberSaveable로 ViewModel 인스턴스 관리
- CardRegistrationScreen에서 ViewModel의 상태와 이벤트를 관찰하여 UI 업데이트
- strings.xml에 유효성 검사 및 상태 메시지 추가 및 수정

moondev03 committed:
build: .gitkeep 삭제

moondev03 committed:
refactor: 컴포넌트 테스트 클래스 패키지 이동 및 이름 변경

- 컴포넌트 테스트 클래스를 `woowacourse.payments.ui` 패키지에서
`woowacourse.payments.ui.component` 패키지로 이동했습니다.

- 또한, `CardExpirationDateTextFieldTest` 클래스에서 오류 메시지 "유효하지 않은 만료일 입니다."를 "유효하지 않은 만료일입니다."로 수정했습니다.

moondev03 committed:
test: DefaultPaymentCardValidator 테스트 코드 추가

moondev03 committed:
test: GroupedVisualTransformation 포맷팅 테스트 추가

moondev03 committed:
style: 카드 소유자 이름 입력창 글자수 구분자 변경

giovannijunseokim commented on app/src/androidTest/java/woowacourse/payments/ui/component/CardExpirationDateTextFieldTest.kt:
```diff
@@ -0,0 +1,107 @@
+            .assertTextEquals("12 / 3")
+    }
+
+    @Test
+    fun `만료일의_연도는_현재_연도_이후로_입력할_수_있다`() {
+        // when
+        composeTestRule
+            .onNodeWithTag(TEST_TAG)
+            .performTextInput("0925")
+
+        // then
+        composeTestRule
+            .onNodeWithTag(TEST_TAG, useUnmergedTree = true)
+            .assertTextEquals("09 / 25")
+    }
```
이 테스트 코드는 내년이 되면 실패할 수 있을 것 같아요. 어떻게 바꿔볼 수 있을까요? 🤔

giovannijunseokim commented on app/src/androidTest/java/woowacourse/payments/ui/component/CardNumberTextFieldTest.kt:
```diff
@@ -0,0 +1,96 @@
+    @Before
+    fun setUp() {
+        composeTestRule.setContent {
+            var state by remember { mutableStateOf("") }
```
`state`라는 변수명을 사용한 이유가 있을까요? 👀

giovannijunseokim commented on app/src/androidTest/java/woowacourse/payments/ui/component/CardNumberTextFieldTest.kt:
```diff
@@ -0,0 +1,96 @@
+
+    @Test
+    fun `카드_번호는_4자리가_될_때_마다_대시_기호가_붙는다`() {
+        val csvSource =
+            arrayOf(
+                "1111,1111",
+                "11111,1111 - 1",
+                "123456789,1234 - 5678 - 9",
+                "1234567890123,1234 - 5678 - 9012 - 3",
+                "1234567890123456,1234 - 5678 - 9012 - 3456",
+            )
+
+        csvSource.forEach { csv ->
+            val (input, expected) = csv.split(",")
```
`Pair`를 사용해볼 수 있을 것 같습니다. 👍

giovannijunseokim commented on app/src/main/java/woowacourse/payments/domain/DefaultPaymentCardValidator.kt:
```diff
@@ -0,0 +1,50 @@
+import java.time.YearMonth
+
+class DefaultPaymentCardValidator : PaymentCardValidator {
+    override fun validateCardNumber(cardNumber: String): PaymentCardValidationResult =
+        when {
+            cardNumber.length != VALID_CARD_NUMBER_LENGTH -> PaymentCardValidationResult.NOT_FILLED
+            cardNumber.length == VALID_CARD_NUMBER_LENGTH -> PaymentCardValidationResult.VALID
+            else -> PaymentCardValidationResult.INVALID
+        }
```
`cardNumber`로 `"abcd".repeat(4)`가 들어오게 되면 `PaymentCardValidationResult.VALID`가 반환될 수 있을 것 같아요. 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/domain/DefaultPaymentCardValidator.kt:
```diff
@@ -0,0 +1,50 @@
+
+        val inputYearMonth =
+            runCatching { YearMonth.of(CENTURY_OFFSET + year, month) }.getOrNull() ?: return false
+        val currentYearMonth = YearMonth.now()
```
domain 객체가 외부 환경에 영향을 받고있는 것 같아요. 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/common/StringResWithParams.kt:
```diff
@@ -0,0 +1,30 @@
+import android.content.res.Resources
+import androidx.annotation.StringRes
+
+class StringResWithParams(
```
이 객체를 설계하신 의도가 궁금해요. 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/CardExpirationDateTextField.kt:
```diff
@@ -0,0 +1,106 @@
+    modifier: Modifier = Modifier,
+    errorMessage: String? = null,
+) {
+    var isValidYearMonth by rememberSaveable { mutableStateOf(true) }
```
이 값도 상태로 존재해야 할까요?
다른 상태의 파생된 형태로 존재할 수는 없을까요? 🤔

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/CardExpirationDateTextField.kt:
```diff
@@ -0,0 +1,106 @@
+            onCardExpirationDateChanged(newExpirationDate)
+        },
+        isError = errorMessage != null,
+        supportingText = {
+            if (isValidYearMonth.not()) {
+                Text(text = stringResource(R.string.card_expiration_date_text_invalid_format))
+                return@OutlinedTextField
+            }
+            errorMessage?.let { message -> Text(text = message) }
+        },
```
얼리 리턴을 잘 활용해 주셨어요! 👍 

하지만 [**컴포저블 함수, 컴포저블 람다 내에서 얼리 리턴을 하는 것은 의도치 않은 부작용을 낳을 수 있다는 의견**](https://medium.com/android-ideas/avoid-return-statements-in-jetpack-compose-9596deb2059f) 또한 존재하는데요. 🤔 

> composables are functions but behave more like classes, it’s easy to make mistakes with early returns, especially as the code inevitably changes over time.

다이스는 이에 대해 어떻게 생각하시는지 궁금해요. 👀

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/CardExpirationDateTextField.kt:
```diff
@@ -0,0 +1,106 @@
+    )
+}
+
+private fun isValidYearMonth(cardExpirationDate: String): Boolean {
+    if (cardExpirationDate.length != CARD_EXPIRATION_DATE_LENGTH) return true
+
+    val month = cardExpirationDate.take(2).toInt()
+    val year = (CENTURY_PREFIX + cardExpirationDate.takeLast(2)).toInt()
+    return runCatching { YearMonth.of(year, month) }.isSuccess
+}
```
코드 맥락상 `cardExpirationDate.length != CARD_EXPIRATION_DATE_LENGTH`가 `true`를 반환하는 이유는 이해가 갑니다.
다만 해당 함수만 본다면 조금 어색하게 느껴질 수 있을 것 같아요. 🤔

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/CardExpirationDateTextField.kt:
```diff
@@ -0,0 +1,106 @@
+
+@Preview(showBackground = true, backgroundColor = 0xFFFFFFFF)
+@Composable
+fun CardExpirationDateTextFieldPreview() {
```
`Preview` 컴포저블 또한 `private`하게 만들 수 있습니다! 👍

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/CardExpirationDateTextField.kt:
```diff
@@ -0,0 +1,106 @@
+@Preview(showBackground = true, backgroundColor = 0xFFFFFFFF)
+@Composable
+fun CardExpirationDateTextFieldPreview() {
+    Column(modifier = Modifier.padding(12.dp)) {
+        CardExpirationDateTextField(
+            cardExpirationDate = "",
+            onCardExpirationDateChanged = {},
+        )
+
+        CardExpirationDateTextField(
+            cardExpirationDate = "1226",
+            onCardExpirationDateChanged = {},
+        )
+
+        CardExpirationDateTextField(
+            cardExpirationDate = "1326",
+            onCardExpirationDateChanged = {},
+            errorMessage = "유효하지 않은 만료일 입니다.",
+        )
+    }
```
[**@PreviewParameter**](https://developer.android.com/develop/ui/compose/tooling/previews#preview-data)를 사용해볼 수 있을 것 같습니다. 👍

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/registration/CardRegistrationScreenUiState.kt:
```diff
@@ -0,0 +1,27 @@
+
+@Parcelize
+data class CardRegistrationScreenUiState(
+    val cardNumber: String = "",
+    val cardNumberValidationResult: PaymentCardValidationResult = PaymentCardValidationResult.NOT_FILLED,
```
`cardNumberValidationResult`같은 경우, `cardNumber`에서 파생된 형태로 존재할 수도 있을 것 같은데요. 별도의 파라미터로 두신 이유가 궁금합니다. 👀

giovannijunseokim reviewed (approved):
리뷰 반영 고생하셨습니다! 🔥 
이번 미션은 여기서 머지할테니 남긴 리뷰 확인해주시면 감사드리겠습니다. 🙇 

다음 미션도 응원하겠습니다! 👍
